shod.1
.Dd September 10, 2022
.Dt SHOD 1
.Os
.Sh NAME
.Nm shod ,
.Nm shodc
.Nd mouse-focused window manager
.Sh SYNOPSIS
.Nm shod
.Op Fl AcdhSstW
.Op Ar file
.Pp
.Nm shodmenu
.Pp
.Nm shodc
.Cm close
.Op Ar win_id
.Nm shodc
.Cm cycle
.Op Fl s
.Nm shodc
.Cm desk
.Nm shodc
.Cm docks
.Op Fl l
.Op Ar win_id ...
.Nm shodc
.Cm exit
.Nm shodc
.Cm focus
.Op Fl clrtbpnLRTBPN
.Op Ar win_id
.Nm shodc
.Cm geom
.Op Fl r
.Op Fl x Ar N
.Op Fl y Ar N
.Op Fl w Ar N
.Op Fl h Ar N
.Op Ar win_id
.Nm shodc
.Cm goto
.Op Fl m Ar mon_id
.Ar desk_id
.Nm shodc
.Cm list
.Op Fl cl
.Op Ar win_id ...
.Nm shodc
.Cm sendto
.Op Fl m Ar mon_id
.Ar desk_id
.Op Ar win_id
.Nm shodc
.Cm state
.Op Fl ATR
.Op Fl abfMmsy
.Op Ar win_id
.Sh DESCRIPTION
.Nm shod
is a multi\-monitor floating reparenting X11 window manager which supports tiled and tabbed containers.
.Nm
works only via mouse with a given key modifier (Alt by default)
and by responding to client messages
(it is needed a controller called
.Nm shodc
to control
.Nm shod Ns ).
.Pp
If a file is given as argument,
.Nm
passes this file to
.Xr sh 1
to run the commands in that file after
.Nm
has initialized itself.
However, if the argument is an hyphen
.Pq "\-" ,
.Nm
calls
.Xr sh 1
to run commands read from the standard input instead.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A
Use
.Qq "Alt"
as modifier key.
More technically, use
.Qq "Alt_L"
as the key to perform container cycling (alt-tab); and use
.Qq "Mod1Mask"
as the modifier for mouse operations.
This option is incompatible with
.Fl W .
.It Fl c
Honor configure requests.
By default,
.Nm
ignore requests to configure a window size or position after the window is mapped.
A client may request the window mager to configure (ie', resize or reposition) its own window.
For example,
.Xr xterm 1
sends a configure request for the window manager to resize its window
when the user changes its font size.
Clients requesting to change its own size may be annoying,
so configure requests are ignored by default;
but they can be honored by invoking
.Nm
with the
.Fl c
command-line option.
.It Fl d
Make dialogs float.
By default, dialog windows are mapped inside the client window.
With this option, dialogs are floating windows instead.
(Technically, this option makes dialog windows behave the same as menu windows).
.It Fl h
Disable setting obscured windows as hidden.
By default, windows whose containers are obscured by other containers are set as
hidden (aka minimized).
This default behavior emulates the way Plan9's rio lists obscured windows as hidden.
With this option set, only manually minimized containers are set as hidden.
.It Fl S
Use sloppy focus rather than click-to-focus between windows in the same container.
.It Fl s
Use sloppy focus rather than click-to-focus between containers.
.It Fl t
Disable container cycling (aka
.Dq "alt-tab" ) .
.It Fl W
Use
.Qq "Super"
(also known as
.Qq "Window Key" ) .
as modifier key.
More technically, use
.Qq "Super_L"
as the key to perform container cycling (alt-tab); and use
.Qq "Mod4Mask"
as the modifier for mouse operations.
This option is incompatible with
.Fl A .
.El
.Pp
.Nm shodmenu
is the command called by
.Nm shod
when the user interacts with the menu button (the button at the top far left of a title bar).
This command should be a script written by the user;
and usually invokes a menu program like
.Xr xmenu 1 .
Before calling
.Nm shodmenu ,
the environment variables
.Ev WINDOWID
and
.Ev WINDOWPOS
(containing the id and position of the window whose menu button was activated, respectively)
are set.
An example
.Nm shodmenu
script is distributed with
.Nm shod .
.Pp
.Nm shodc
is the remote controller for
.Nm shod .
Its first argument must be an operation to be performed.
The following arguments are the options for the operation
(each operation accepts a different set of options).
The last arguments are the operation's arguments, such as a window ID or a desktop ID.
.Pp
Known operations for
.Nm shodc
are listed below.
.Ss Close window
The
.Cm close
operation closes a window whose ID is provided as argument.
If no argument is provided, close the active window.
.Ss Cycle containers
The
.Cm cycle
operation initiates the cycling through the open containers on the current desktop.
This cycling is generally called
.Qq "alt-tab" .
Cycling is performed as
.Qq "Tab"
is pressed, and terminates when the modifier key is released.
.Ss List desktops
The
.Cm desks
operation lists the desktops, one per line.
.Pp
If the line begins with an asterisk, the desktop is the focused one;
If the line begins with an hyphen, the desktop has an urgent window in it.
The number is the number of windows in the desktop.
.Ss Exit shod
The
.Cm exit
operation exits the
.Nm shod
window manager.
.Ss Focus window
The
.Cm focus
operation focus a window whose ID is provided as argument.
If an option is provided, focus a window relative to the window provided as argument.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
Cycle focus.
This option only makes sense when using
.Fl n ,
.Fl p ,
.Fl N ,
or
.Fl P .
.It Fl l
Focus window in the closest container to the left of the container of the window provided as argument
(or the active window, if no argument is provided).
.It Fl r
Focus window in the closest container to the right of the container of the window provided as argument
(or the active window, if no argument is provided).
.It Fl t
Focus window in the closest container on the top of the container of the window provided as argument
(or the active window, if no argument is provided).
.It Fl b
Focus window in the closest container on the bottom of the container of the window provided as argument
(or the active window, if no argument is provided).
.It Fl p
Focus window in the previous container in the focus list
in relation to the container of the window provided as argument
(or the active window, if no argument is provided).
.It Fl n
Focus window in the next container in the focus list
in relation to the container of the window provided as argument
(or the active window, if no argument is provided).
.It Fl L
Focus window in the column to the left of the window provided as argument
(or the active window, if no argument is provided).
.It Fl R
Focus window in the column to the right of the window provided as argument
(or the active window, if no argument is provided).
.It Fl T
Focus window in the row above (on the top of) the window provided as argument
(or the active window, if no argument is provided).
.It Fl B
Focus window in the row below (on the bottom of) the window provided as argument
(or the active window, if no argument is provided).
.It Fl P
Focus window in the tab previous to the window provided as argument
(or the active window, if no argument is provided).
.It Fl N
Focus window in the tab next to the window provided as argument
(or the active window, if no argument is provided).
.El
.Ss Set geometry
The
.Cm geom
operation sets the geometry (position and size) of the container of the window whose ID is provided as argument.
If no argument is provided, sets the geometry of the container of the active window.
If no position (set by the options
.Fl x
and
.Fl y )
is provided, move window to position 0,0 (top left corner).
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl r
Relative.
All position and size values are relative to the container's current position and size.
.It Fl x Ar N
Set the position on the X axis to N.
.It Fl y Ar N
Set the position on the Y axis to N.
.It Fl w Ar N
Set the width of the container to N.
.It Fl h Ar N
Set the height of the container to N.
.El
.Ss Go to desktop
The
.Cm goto
operation goes to the desktop ID provided as argument.
Different of other window managers, shod counts desktop from 1;
So the first desktop is the desktop 1, not the desktop 0.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl M Ar monitor
Goes to a desktop on the provided monitor rather than on the currently focused monitor.
Monitors are counted from 1, not from 0.
.El
.Ss List windows
The
.Cm list
operation lists windows managed by
.Nm shod ,
one entry per line.
If one or more window ID are provided as arguments, list only those windows.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
List only the focused window of each container.
.It Fl l
Long list format.
More information on this format below.
.El
.Pp
If the
.Fl l
option is given, the following information (delimited by tabs)
is displayed for each window, in the following order:
.Bl -enum -compact
.It
Window state.
.It
Window desktop number.
.It
Window geometry (size and position).
.It
ID of the container the window is on.
.It
ID of the group the window is on.
.It
ID of the window.
.It
Name of the window.
.El
.Pp
The window state consists of a sequence of eight characters,
each one meaning a state for the container.
If a character is an hyphen
.Pq \-
the state is not set or does not apply to the window.
The state characters, in the order they appear, are the following:
.Bl -enum -compact
.It
An
.Dq Sy y
indicates that the window's container is sticky.
.It
An
.Dq Sy M
indicates that the window's container is maximized.
.It
An
.Dq Sy m
indicates that the window's container is minimized.
.It
An
.Dq Sy f
indicates that the window's container is fullscreen.
.It
An
.Dq Sy s
indicates that the window's container is shaded.
.It
An
.Dq Sy a
indicates that the window's container is above others.
An
.Dq Sy b
indicates that the window's container is below others.
.It
An
.Dq Sy u
indicates that the window has the urgency hint set.
An
.Dq Sy a
indicates that the window demands attention.
An
.Dq Sy U
indicates that the window is both urgent and demands attention.
.It
An
.Dq Sy a
indicates that the window is active.
An
.Dq Sy f
indicates that the window is focused.
An
.Dq Sy F
indicates that the window is both active and focused.
.El
.Ss List docks
The
.Cm docks
operation lists docks and bars managed by
.Nm shod ,
one entry per line.
If one or more window ID are provided as arguments, list only those windows.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl l
Long list format.
More information on this format below.
.El
.Pp
If the
.Fl l
option is given, the following information (delimited by tabs) is displayed for each window,
in the following order:
.Bl -enum -compact
.It
Window state.
.It
Window geometry.
.It
ID of the window.
.It
Name of the window.
.El
.Ss Send to desktop
The
.Cm sendto
operation sends to the desktop ID provided as first argument
the container of the window whose ID is provided as second argument.
If no window ID is provided, sends the container of the active window to that desktop.
Different of other window managers, shod counts desktop from 1;
so the first desktop is the desktop 1, not the desktop 0.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl M Ar monitor
Sends to a desktop on the provided monitor rather than on the currently focused monitor.
Monitors are counted from 1, not from 0.
.El
.Ss Set window state
The
.Cm state
operation sets the state of a window whose ID is provided as argument.
If no argument is provided, sets the state of the active window.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
Above state.
.It Fl b
Below state.
.It Fl f
Fullscreen state.
.It Fl M
Maximized state.
.It Fl m
Minimized state.
.It Fl s
Shaded state.
.It Fl y
Sticky state.
.It Fl A
Add state (force state to be set).
.It Fl T
Toggle state (set state if it is unset, or unset it if it is set).
This is the default.
.It Fl R
Remove state (force state to be unset).
.El
.Sh USAGE
.Nm shod
maintains one virtual monitor for each physical monitor found by
.Xr Xrandr 3 .
One of the monitors is the focused one, where new windows go to when they are created.
Each monitor contains a different set of virtual desktops (or "desktop" for short).
One of the desktops of a monitor is the focused desktop for that monitor.
.Pp
Each monitor has an area called
.Dq "container area" ,
within containers are spawned and can be maximized.
The size and position of a monitor's container area can be changed by bars and the dock.
.Pp
Most client windows are displayed in containers;
but some windows are special and are displayed in different ways.
.Ss Containers
Containers are floating windows.
A container contains sub-windows (called tiles) organized in columns, rows, and tabs;
each tile represents a client and can have dialog windows and menu windows associated with it.
A new container is created in the focused desktop of the focused monitor;
this new container is placed in an empty area of the screen.
.Pp
A container has borders
(four edge handles and four corner handles) around it;
borders are always visible, except when the container is fullscreen.
Borders are used to move and resize containers (more information on that below).
A container has divisors between its columns, and between the rows within each column.
Divisos are used to resize the tiles by dragging them with the first mouse button.
A container has title bars above each row.
Title bars contains tabs, the left (stack) button, and the right (close) button.
More information on those below.
.Pp
Keyboard input goes to the focused client of the focused container.
A container can be focused by clicking on it with the 1st mouse button;
the click is passed to the application, which can process it in a particular way.
The focused container is decorated with a visually distinct decoration
(blue in the default theme);
while other containers have either the common decoration
(gray in the default theme),
or have the urgent decoration
(red in the default theme).
.Pp
Containers are stacked one above the other in the virtual Z axis.
The position of the container in this Z axis can be changed by an operation called
.Dq raising .
A container can be raised by clicking on it with the 1st mouse button.
The list of containers in this Z axis, from the one in the bottom to the topmost, is called the
.Dq "stacking order" .
The stacking order is organized in four layers.
When a container is raised, it move to the top of its layers.
The layers are the following (from bottomost to topmost):
.Bl -enum -compact
.It
The bottom layer (for containers below others).
.It
The middle layer (for normal containers).
.It
The upper layer (for containers above others).
.It
The fullscreen layer (for fullscreen containers).
.El
.Pp
Each container can have one or more of the following states
(which can be set by
.Nm shodc
.Cm state ) .
.Bl -tag -width Ds
.It Maximized
The container occupies the entire container area.
.It Fullscreen
The container occupies the entire monitor and its borders are hidden.
.It Minimized
The container is not shown on the monitor.
.It Shaded
The container is resized to show only the title bars.
.It Sticky
The container is stickied to the monitor,
and appears on the screen no matter which desktop is selected.
.It Above/Below
Those two states are mutually exclusive.
The container is raised above or lowered below other containers.
.El
.Pp
A non-fullscreen, non-maximized container can be moved by the following methods:
.Bl -bullet -compact
.It
By dragging the container border with the third mouse button.
.It
By dragging a title bar with the first mouse button.
Or
.It
By pressing the modifier key and dragging any part of the container with the first mouse button.
.El
.Pp
Each title-bar has tabs, which displayes the title of its client.
Dragging a tab with the first mouse button moves the entire container.
Dragging a tab with the third mouse button detaches the tab from the container.
A detached tab, while being dragged, can be reattached into another container
(or the same container) by dropping it on the title bar, border or divisor;
or can be made into a new container by dropping it elsewhere.
Double-clicking a tab (un)stretches its row.
Scrolling a tab up shades its container.
Scrolling a down up unshades its container.
.Pp
Moving a container to the top of the screen maximizes it.
.Pp
Each title bar has a left button.
Clicking on the left title-bar button with the first mouse button stretches the column
by maximizing its rows (and minimizing the other rows in the same column),
or undoes this state.
.Pp
Each title bar has a right button.
Clicking on the right title-bar button with the first mouse button
closes the focused client or its top dialog.
.Pp
Containers can be cycled using the modifier key set with either the
.Fl A
or
.Fl W
command-line option
.Po
.Cm Alt_L
by default
.Pc
followed by the
.Cm Tab
key.
The
.Cm Tab
key can be further modified by
.Cm Shift
to cycle in the oposite direction.
This mechanism is usually called
.Dq "alt-tab"
because of the key combination that usually triggers it.
This mechanism can be turned off by invoking
.Nm shod
with the
.Fl t
command-line option.
.Ss Dialog
Windows that are transient for another managed windows (called its leader)
are mapped in the center of the leader.
.Pp
Dialogs are small windows that communicates information to the user
or prompts for a response.
.Pp
.Nm shod
only changes the position and size of a dialog window when the size of its leader changes.
.Ss Splash screens
Windows of type
.Ic _NET_WM_WINDOW_TYPE_SPLASH
(called splash screens)
are mapped above all other windows and are stacked on the order they are spawned.
Splash screens cannot be manipulated.
Splash screens have no decoration around them.
.Pp
Splash screens are transient windows that appear temporarily while an application is loading.
.Pp
.Nm shod
centers the splash screen on the monitor.
.Ss Menus
windows of type
.Ic _NET_WM_WINDOW_TYPE_MENU ,
.Ic _NET_WM_WINDOW_TYPE_UTILITY ,
.Ic _NET_WM_WINDOW_TYPE_TOOLBAR ,
or
.Ic MWM_TEAROFF_WINDOW
(called menu windows)
are windows that cannot be tiled or tabbed into a container
and are optionally tied to a leader window.
They are floating windows that always apear on top of their leader
and are not listed on the list of clients.
If a menu has no leader window, they apear on top of all windows.
.Pp
Menu windows, often called
.Dq "torn off windows" ,
are pinnable menus, utility windows, and toolbar windows
(ie', toolbar and menus "torn off" from the main application).
.Pp
The user can change the position of a menu window in the same way of
changing the position of a container.
.Ss Prompt
A window of type
.Ic _NET_WM_WINDOW_TYPE_PROMPT
(called prompt window)
is mapped on the top of the focused monitor.
This window will stay focused and mapped until be closed,
or a mouse button is pressed outside that window.
This is an EWMH extention, only used by
.Xr xprompt 1 .
.Pp
.Nm shod
does not change the size of the prompt window.
However, shod changes its position.
.Ss Desktop windows
Windows of type
.Ic _NET_WM_WINDOW_TYPE_DESKTOP
(called desktop windows)
are mapped below all other windows and are stacked on the order they are spawned.
Desktop windows cannot be manipulated.
Desktop windows have no decoration around them.
.Pp
Desktop windows indicates a desktop feature.
That includes windows such as
.Xr conky 1
and windows that manage desktop icons.
.Pp
.Nm shod
does not change the size nor the position of desktop windows.
.Ss Notifications
Windows of type
.Ic _NET_WM_WINDOW_TYPE_NOTIFICATION
(called notifications)
are popped up on the top right corner, one above another.
Notification windows cannot be manipulated.
Notification windows have a decoration around them;
this decoration is the same as the borders of the active container
(or, for urgent notifications, the same as the borders of an urgent container).
.Pp
An example of a notification window would be a bubble appearing with
informative text such as
.Dq "Your laptop is running out of power"
etc.
.Pp
.Nm shod
can change the size and the position of notification windows.
.Ss Docks
Shod's internal dock and windows of type
.Ic _NET_WM_WINDOW_TYPE_DOCK
(called panels, bars, or external docks)
are mapped on a side of a monitor and are called
.Em docks .
Dock windows cannot be manipulated,
have no decoration and do not receive input focus.
.Pp
Depending on ther maximized status,
docks can change the size of a region of the monitor called
.Dq "container area" .
The container area is the region of the monitor that a maximized container occupies.
The container area is also the region of the monitor inside which containers are spawned.
.Pp
Each dock can have one or more of the following states
(which can be set by
.Nm shodc
.Cm state ) .
.Bl -tag -width Ds
.It Maximized
The dock shrinks the container area.
.It Minimized
The dock is hidden.
.It Above/Below
Those two states are mutually exclusive.
The dock is placed above containers' upper layer, or below containers' bottom layer, respectively.
.El
.Pp
Examples of docks are a taskbar that shows which programs are currently running,
or a status bar that shows information about the system.
.Pp
.Nm shod
does not change the size nor the position of bar windows.
.Ss Dockapps
Windows that initiate in the
.Ic WithdrawnState
(called dockapps)
are mapped inside Shod's internal dock.
The dock is a panel or bar that appears on the edge of the first monitor.
Inside the dock, dockapps are organized by order of map request.
.Pp
Dockapps, or docked applications are windows which appear to reside
inside an icon or a dock rather than a container.
.Pp
Dockapps can have one of three possible states: spaced, shrunk, or extended.
The state of a dockapp can be set with the
.Ic "CLASS.NAME.ROLE.state"
X resource.
.Pp
A
.Dq Em "spaced"
dockapp is centered on a slot whose size is multiple of, by default, 64 pixels
(this value can be changed with the
.Ic "dockSpace"
X resource).
So for example, by default, if a dockapp has 58 pixels, it is centered on a slot of 64 pixels (64 * 1);
but if a dockapp has 100 pixels, it is centered on a slot of 128 pixels (64 * 2).
This is the default state for a dockapp.
.Pp
A
.Dq Em "shrunk"
dockapp
has its slot resized to the size of the dockapp itself.
.Pp
An
.Dq Em "extended"
dockapp is resized to fit all the remaining unused space of the dock.
For example, if the only content of the dock is an extended taskbar dockapp,
this dockapp is resized to fit all the dock.
.Pp
an

shod.c
#include <sys/wait.h>

#include <err.h>
#include <limits.h>
#include <locale.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <X11/extensions/Xrandr.h>

#include "shod.h"

#define DOCK_TITLE      "shod's dock"
#define WMNAME          "shod"

static int (*xerrorxlib)(Display *, XErrorEvent *);

/* shared variables */
unsigned long clientmask = CWEventMask | CWColormap | CWBackPixel | CWBorderPixel;
XSetWindowAttributes clientswa = {
	.event_mask = StructureNotifyMask | SubstructureRedirectMask
	            | ButtonReleaseMask | ButtonPressMask
	            | FocusChangeMask | Button1MotionMask
};
struct WM wm = { .running = 1 };
struct Dock dock;

/* show usage and exit */
static void
usage(void)
{
	(void)fprintf(stderr, "usage: shod [-AcdhSstW] [file]\n");
	exit(1);
}

/* call fork checking for error; exit on error */
static pid_t
efork(void)
{
	pid_t pid;

	if ((pid = fork()) < 0)
		err(1, "fork");
	return pid;
}

/* call execlp checking for error; exit on error */
static void
execshell(char *filename)
{
	char *argv[3];

	if ((argv[0] = getenv(SHELL)) == NULL)
		argv[0] = DEF_SHELL;
	if (filename[0] == '-' && filename[1] == '\0')
		argv[1] = NULL;         /* read commands from stdin */
	else
		argv[1] = filename;     /* read commands from file */
	argv[2] = NULL;
	if (execvp(argv[0], argv) == -1) {
		err(1, "%s", argv[0]);
	}
}

/* error handler */
static int
xerror(Display *dpy, XErrorEvent *e)
{
	/* stolen from berry, which stole from katriawm, which stole from dwm lol */

	/* There's no way to check accesses to destroyed windows, thus those
	 * cases are ignored (especially on UnmapNotify's). Other types of
	 * errors call Xlibs default error handler, which may call exit. */
	if (e->error_code == BadWindow ||
	    (e->request_code == X_SetInputFocus && e->error_code == BadMatch) ||
	    (e->request_code == X_PolyText8 && e->error_code == BadDrawable) ||
	    (e->request_code == X_PolyFillRectangle && e->error_code == BadDrawable) ||
	    (e->request_code == X_PolySegment && e->error_code == BadDrawable) ||
	    (e->request_code == X_ConfigureWindow && e->error_code == BadMatch) ||
	    (e->request_code == X_ConfigureWindow && e->error_code == BadValue) ||
	    (e->request_code == X_GrabButton && e->error_code == BadAccess) ||
	    (e->request_code == X_GrabKey && e->error_code == BadAccess) ||
	    (e->request_code == X_CopyArea && e->error_code == BadDrawable) ||
	    (e->request_code == 139 && e->error_code == BadDrawable) ||
	    (e->request_code == 139 && e->error_code == 143))
		return 0;

	fprintf(stderr, "shod: ");
	return xerrorxlib(dpy, e);
	exit(1);        /* unreached */
}

/* startup error handler to check if another window manager is already running */
static int
xerrorstart(Display *dpy, XErrorEvent *e)
{
	(void)dpy;
	(void)e;
	errx(1, "another window manager is already running");
}

/* read command-line options */
static char *
getoptions(int argc, char *argv[])
{
	int c;

	while ((c = getopt(argc, argv, "AcdhSstW")) != -1) {
		switch (c) {
		case 'A' :
			config.altkeysym = XK_Alt_L;
			config.modifier = Mod1Mask;
			break;
		case 'c':
			config.honorconfig = 1;
			break;
		case 'd':
			config.floatdialog = 1;
			break;
		case 'h':
			config.disablehidden = 1;
			break;
		case 'S':
			config.sloppytiles = 1;
			break;
		case 's':
			config.sloppyfocus = 1;
			break;
		case 't':
			config.disablealttab = 1;
			break;
		case 'W' :
			config.altkeysym = XK_Super_L;
			config.modifier = Mod4Mask;
			break;
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc > 1)
		usage();
	return *argv;
}

/* initialize signals */
static void
initsignal(void)
{
	struct sigaction sa;

	/* remove zombies, we may inherit children when exec'ing shod in .xinitrc */
	sa.sa_handler = SIG_IGN;
	sa.sa_flags = SA_NOCLDSTOP | SA_NOCLDWAIT | SA_RESTART;
	sigemptyset(&sa.sa_mask);
	if (sigaction(SIGCHLD, &sa, NULL) == -1)
		err(1, "sigaction");
	while (waitpid(-1, NULL, WNOHANG) > 0)
		;
}

/* initialize cursors */
static void
initcursors(void)
{
	wm.cursors[CURSOR_NORMAL] = XCreateFontCursor(dpy, XC_left_ptr);
	wm.cursors[CURSOR_MOVE] = XCreateFontCursor(dpy, XC_fleur);
	wm.cursors[CURSOR_NW] = XCreateFontCursor(dpy, XC_top_left_corner);
	wm.cursors[CURSOR_NE] = XCreateFontCursor(dpy, XC_top_right_corner);
	wm.cursors[CURSOR_SW] = XCreateFontCursor(dpy, XC_bottom_left_corner);
	wm.cursors[CURSOR_SE] = XCreateFontCursor(dpy, XC_bottom_right_corner);
	wm.cursors[CURSOR_N] = XCreateFontCursor(dpy, XC_top_side);
	wm.cursors[CURSOR_S] = XCreateFontCursor(dpy, XC_bottom_side);
	wm.cursors[CURSOR_W] = XCreateFontCursor(dpy, XC_left_side);
	wm.cursors[CURSOR_E] = XCreateFontCursor(dpy, XC_right_side);
	wm.cursors[CURSOR_V] = XCreateFontCursor(dpy, XC_sb_v_double_arrow);
	wm.cursors[CURSOR_H] = XCreateFontCursor(dpy, XC_sb_h_double_arrow);
	wm.cursors[CURSOR_HAND] = XCreateFontCursor(dpy, XC_hand2);
	wm.cursors[CURSOR_PIRATE] = XCreateFontCursor(dpy, XC_pirate);
}

static void
initxrm(void)
{
	static struct {
		const char *class, *name;
	} resourceids[NRESOURCES] = {
#define X(res, s1, s2) [res] = { .class = s1, .name = s2, },
		RESOURCES
#undef  X
	};
	long n;
	int i;
	char *value;

	XrmInitialize();
	wm.application.class = XrmPermStringToQuark("Shod");
	wm.application.name = XrmPermStringToQuark("shod");
	wm.anyresource = XrmPermStringToQuark("?");
	for (i = 0; i < NRESOURCES; i++) {
		wm.resources[i].class = XrmPermStringToQuark(resourceids[i].class);
		wm.resources[i].name = XrmPermStringToQuark(resourceids[i].name);
	}
	if (!settheme())
		exit(EXIT_FAILURE);
	setresources(XResourceManagerString(dpy));
	if (xdb != NULL) {
		value = getresource(
			xdb,
			(XrmClass[]){
				wm.application.class,
				wm.resources[RES_NDESKTOPS].class,
				NULLQUARK,
			},
			(XrmName[]){
				wm.application.name,
				wm.resources[RES_NDESKTOPS].name,
				NULLQUARK,
			}
		);
		if (value != NULL && (n = strtol(value, NULL, 10)) > 0 && n < 100) {
			config.ndesktops = n;
		}
	}
}

/* create dock window */
static void
initdock(void)
{
	XSetWindowAttributes swa;

	TAILQ_INIT(&dock.dappq);
	dock.pix = None;
	swa.event_mask = SubstructureNotifyMask | SubstructureRedirectMask;
	swa.background_pixel = BlackPixel(dpy, screen);
	swa.border_pixel = BlackPixel(dpy, screen);
	swa.colormap = colormap;
	dock.obj.win = XCreateWindow(
		dpy, root,
		0, 0, 1, 1, 0,
		depth, InputOutput, visual,
		clientmask, &swa
	);
	dock.state = MAXIMIZED;
	dock.obj.class = dock_class;
	settitle(dock.obj.win, DOCK_TITLE);
}

/* create dummy windows used for controlling focus and the layer of clients */
static void
initdummywindows(void)
{
	int i;
	XSetWindowAttributes swa;

	for (i = 0; i < LAYER_LAST; i++) {
		wm.layers[i].ncols = 0;
		wm.layers[i].frame = XCreateSimpleWindow(dpy, root, 0, 0, 1, 1, 0, 0, 0);
		XRaiseWindow(dpy, wm.layers[i].frame);
		TAILQ_INSERT_HEAD(&wm.stackq, &wm.layers[i], raiseentry);
	}
	swa = clientswa;
	swa.event_mask |= KeyPressMask;
	wm.checkwin = wm.focuswin = wm.dragwin = wm.restackwin = XCreateWindow(
		dpy, root,
		- (2 * config.borderwidth + config.titlewidth),
		- (2 * config.borderwidth + config.titlewidth),
		2 * config.borderwidth + config.titlewidth,
		2 * config.borderwidth + config.titlewidth,
		0, depth, CopyFromParent, visual,
		clientmask,
		&swa
	);
}

/* map and hide dummy windows */
static void
mapdummywins(void)
{
	XMapWindow(dpy, wm.focuswin);
	XSetInputFocus(dpy, wm.focuswin, RevertToParent, CurrentTime);
}

/* run stdin on sh */
static void
autostart(char *filename)
{
	pid_t pid;

	if (filename == NULL)
		return;
	if ((pid = efork()) == 0) {
		if (efork() == 0)
			execshell(filename);
		exit(0);
	}
	waitpid(pid, NULL, 0);
}

static void
checkotherwm(void)
{
	/*
	 * NOTE: Do we need to select SubstructureNotifyMask on the root window?
	 *
	 * We will always select StructureNotifyMask on the client windows we
	 * are requested to map, so selecting StructureNotifyMask on the root
	 * window seems redundant.
	 *
	 * I have removed this mask in the bitmask passed as the third parameter
	 * to XSelectInput(3) down here.  If anything ever brake, just add it
	 * back.
	 */
	xerrorxlib = XSetErrorHandler(xerrorstart);
	XSelectInput(
		dpy, root,
		SubstructureRedirectMask |      /* so clients request us to map */
		StructureNotifyMask |           /* get changes on root configuration */
		PropertyChangeMask |            /* get changes on root properties */
		ButtonPressMask                 /* to change monitor when clicking */
	);
	XSync(dpy, False);
	(void)XSetErrorHandler(xerror);
	XSync(dpy, False);
}

/* destroy dummy windows */
static void
cleandummywindows(void)
{
	int i;

	XDestroyWindow(dpy, wm.checkwin);
	for (i = 0; i < LAYER_LAST; i++) {
		XDestroyWindow(dpy, wm.layers[i].frame);
	}
}

/* free cursors */
static void
cleancursors(void)
{
	size_t i;

	for (i = 0; i < CURSOR_LAST; i++) {
		XFreeCursor(dpy, wm.cursors[i]);
	}
}

/* clean window manager structures */
static void
cleanwm(void)
{
	struct Monitor *mon;
	struct Object *obj;
	struct Container *c;

	while ((c = TAILQ_FIRST(&wm.focusq)) != NULL)
		containerdel(c);
	while ((obj = TAILQ_FIRST(&wm.notifq)) != NULL)
		(*obj->class->unmanage)(obj);
	while ((obj = TAILQ_FIRST(&wm.barq)) != NULL)
		(*obj->class->unmanage)(obj);
	while ((obj = TAILQ_FIRST(&wm.splashq)) != NULL)
		(*obj->class->unmanage)(obj);
	while ((obj = TAILQ_FIRST(&dock.dappq)) != NULL)
		(*obj->class->unmanage)(obj);
	while ((mon = TAILQ_FIRST(&wm.monq)) != NULL)
		mondel(mon);
	if (dock.pix != None)
		XFreePixmap(dpy, dock.pix);
	XDestroyWindow(dpy, dock.obj.win);
}

/* shod window manager */
int
main(int argc, char *argv[])
{
	XEvent ev;
	char *filename, *wmname;

	if (!setlocale(LC_ALL, "") || !XSupportsLocale())
		warnx("warning: no locale support");
	xinit();
	checkotherwm();
	moninit();
	xinitvisual();
	clientswa.colormap = colormap;
	clientswa.border_pixel = BlackPixel(dpy, screen);
	clientswa.background_pixel = BlackPixel(dpy, screen);

	/* get configuration */
	if ((wmname = strrchr(argv[0], '/')) != NULL)
		wmname++;
	else if (argv[0] != NULL && argv[0][0] != '\0')
		wmname = argv[0];
	else
		wmname = WMNAME;
	filename = getoptions(argc, argv);

	/* check sloppy focus */
	if (config.sloppyfocus || config.sloppytiles)
		clientswa.event_mask |= EnterWindowMask;

	/* initialize queues */
	TAILQ_INIT(&wm.monq);
	TAILQ_INIT(&wm.barq);
	TAILQ_INIT(&wm.splashq);
	TAILQ_INIT(&wm.notifq);
	TAILQ_INIT(&wm.focusq);
	TAILQ_INIT(&wm.stackq);

	/* initialize */
	initsignal();
	initcursors();
	initatoms();
	initdummywindows();
	initdock();
	initxrm();

	/* set up list of monitors */
	monupdate();
	wm.selmon = TAILQ_FIRST(&wm.monq);

	/* initialize ewmh hints */
	ewmhinit(wmname);
	ewmhsetcurrentdesktop(0);
	ewmhsetshowingdesktop(0);
	ewmhsetclients();
	ewmhsetactivewindow(None);

	/* run sh script */
	autostart(filename);

	/* scan windows */
	scan();
	mapdummywins();

	/* change default cursor */
	XDefineCursor(dpy, root, wm.cursors[CURSOR_NORMAL]);

	/* Set focus to root window */
	XSetInputFocus(dpy, root, RevertToParent, CurrentTime);

	/* set modifier key and grab alt key */
	setmod();

	/* run main event loop */
	while (wm.running) {
		(void)XNextEvent(dpy, &ev);
		if (ev.type < LASTEvent && xevents[ev.type] != NULL)
			(*xevents[ev.type])(&ev);
		else if (wm.xrandr && ev.type == wm.xrandrev)
			monevent(&ev);
		if (wm.setclientlist)
			ewmhsetclients();
		wm.setclientlist = false;
	}

	/* clean up */
	cleandummywindows();
	cleancursors();
	cleanwm();
	cleantheme();

	/* clear ewmh hints */
	ewmhsetclients();
	ewmhsetactivewindow(None);

	/* close connection to server */
	XUngrabPointer(dpy, CurrentTime);
	XUngrabKeyboard(dpy, CurrentTime);
	XCloseDisplay(dpy);

	return 0;
}
